### PR TITLE
[tests] Re-enable unittesting and partial fix for the existing errors

### PR DESCRIPTION
**Pull request by Menooker**
The unit-testing is accidentally disabled since the last refactoring of the `test` folder. We need to add a dummy folder to hold `lig.cfg.py` for the C++ based unit tests.

This reveals 2 errors in the current repo.
 * memref-to-cpuruntime should be a function pass in the CPUPipeline (this issue is fixed)
 * unit test failure on `TargetDescriptionAnalysis.CPUMissingValue` (**need to fix** in a follow-up PR @zhczhong  )

Also fixed an error which breaks g++9, when we need a workaround on the CXX_FLAGS.

### CODE REVIEW

**Comment by ZhennanQin** [comment]
Thank you @Menooker !

File: CMakeLists.txt [hunk 1]
```diff
@@ -29,8 +29,11 @@ list(APPEND CMAKE_MODULE_PATH "${PROJECT_SOURCE_DIR}/cmake")
 # https://gcc.gnu.org/bugzilla/show_bug.cgi?id=85827 for details. The issue is
 # fixed in GCC 10.
 if(CMAKE_CXX_COMPILER_ID MATCHES "GNU" AND CMAKE_CXX_COMPILER_VERSION VERSION_LESS "10.0")
+  include(CheckCXXCompilerFlag)
   check_cxx_compiler_flag("-Wno-unused-but-set-parameter" CXX_SUPPORTS_WNO_UNUSED_BUT_SET_PARAMETER)
-  append_if(CXX_SUPPORTS_WNO_UNUSED_BUT_SET_PARAMETER "-Wno-unused-but-set-parameter" CMAKE_CXX_FLAGS)
+  if(CXX_SUPPORTS_WNO_UNUSED_BUT_SET_PARAMETER)
```
Comment (Menooker) [inline CMakeLists.txt:34]: These lines originally occur after including LLVM cmake, so we don't need to include them by ourselves. A recent refactor swaps the location of these lines with include(LLVM...), so it breaks cmake when g++ < 10.0

File: test/mlir/unittests/Analysis/TargetDescriptionAnalysisTest.cpp [hunk 2]
```diff
@@ -62,7 +62,7 @@ dlti.target_system_spec = #dlti.target_system_spec<
 >} {}
 )mlir";
 
-TEST(TargetDescriptionAnalysis, CPUMissingValue) {
+TEST(DISABLED_TargetDescriptionAnalysis, CPUMissingValue) {
```
Comment (Menooker) [inline test/mlir/unittests/Analysis/TargetDescriptionAnalysisTest.cpp:65]: We need to re-enable this UT after it is fixed.

**Review by crazydemo — APPROVED** [review]

**Review by zhczhong — APPROVED** [review]

**Comment by ciyongch** [comment]
So this will fix the escape of unittest from CI/CD, right?

**Comment by Menooker** [comment]
> So this will fix the escape of unittest from CI/CD, right?

Yes. It will!

File: test/mlir/test/gc/Unit/lit.cfg.py [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+# if none of these are present.
+if "TMP" in os.environ:
+    config.environment["TMP"] = os.environ["TMP"]
+if "TEMP" in os.environ:
+    config.environment["TEMP"] = os.environ["TEMP"]
+
+# Propagate HOME as it can be used to override incorrect homedir in passwd
+# that causes the tests to fail.
+if "HOME" in os.environ:
+    config.environment["HOME"] = os.environ["HOME"]
```
Comment (ciyongch) [inline test/mlir/test/gc/Unit/lit.cfg.py:34]: nit: seems we can also put these into var list below？

File: test/mlir/unittests/Analysis/TargetDescriptionAnalysisTest.cpp (hunk 2 above)
Comment (ciyongch) [reply test/mlir/unittests/Analysis/TargetDescriptionAnalysisTest.cpp:65]: Please add a TODO/FIXME for future reminder

File: lib/gc/Transforms/Pipeline.cpp [hunk 4]
```diff
@@ -128,7 +128,7 @@ void populateCPURuntimePasses(mlir::OpPassManager &pm) {
   pm.addPass(createForallToParallelLoopPass());
   pm.addPass(createParallelLoopFusionPass());
   pm.addPass(createLoopInvariantCodeMotionPass());
-  pm.addPass(createConvertMemRefToCPURuntime());
+  pm.addNestedPass<func::FuncOp>(createConvertMemRefToCPURuntime());
```
Comment (xurui1995) [inline lib/gc/Transforms/Pipeline.cpp:131]: can we merge this now?  bench tool has an issue for some mlir, and it needs this change.

**Pull request merged**